xarray/core/dataset.py

```diff
@@ -2366,6 +2339,48 @@ def unstack(self, dim):
 
         return self._replace_vars_and_dims(variables, coord_names)
 
+    def unstack(self, dim=None):
+        """
+        Unstack existing dimensions corresponding to MultiIndexes into
+        multiple new dimensions.
+
+        New dimensions will be added at the end.
+
+        Parameters
+        ----------
+        dim : str or sequence of str, optional
+            Dimension(s) over which to unstack. By default unstacks all
+            MultiIndexes.
+
+        Returns
+        -------
+        unstacked : Dataset
+            Dataset with unstacked data.
+
+        See also
+        --------
+        Dataset.stack
+        """
+        if isinstance(dim, basestring):
+            dims = [dim]
+        elif dim is None:
+            dims = self.dims
+        else:
+            dims = dim
+
+        missing_dims = [d for d in dims if d not in self.dims]
+        if missing_dims:
+            raise ValueError('Dataset does not contain the dimensions: %s'
+                             % missing_dims)
+
+        multi_dims = [d for d in dims if isinstance(self.get_index(d),
+                                                    pd.MultiIndex)]
+
+        result = self
```

Comment:
This is a minor point, but please make a shallow copy rather than ever returning the original object, e.g., `self.copy(deep=False)`. We try to do this consistently in xarray operations so users aren't surprised when they mutate objects.

Example of why this can be problematic:
```python
def unstack_with_history(xarray_obj):
    # this looks like a pure function, but it can actually modify the
    # input object if there's nothing to unstack
    unstacked = xarray_obj.unstack()
    unstacked.attrs['history'] = 'This was unstacked!'
    return unstacked
```

Reply:
Thanks for the explanation. That makes a lot of sense.